### PR TITLE
ci: do dry-run releases in CI

### PR DESCRIPTION
**Pull request by apoelstra**
Will set this to "allow failure" since I think it'll fail during large parts of development, but this will be useful to avoid last-minute problems with releases.

### CODE REVIEW

**Review by tcharding — APPROVED** [review]
ACK 82beb6336c58848a1de439f5f4ce79a8da8a1d65

**Comment by Kixunil** [comment]
Won't this break down when we have changes that involve multiple crates?

**Comment by apoelstra** [comment]
Yes, which is why I will set these all to "allow failure". When we do an actual release we can make sure that the relevant green checkmarks appear on release PRs. (And prior to releases, we will notice if we mess up Cargo.toml metadata.)

**Review by Kixunil — DISMISSED** [review]
ACK 17cd9461996e2b897d39152c85c79d6321e5a9d9

**Comment by tcharding** [comment]
Hah you did a classic Tobin fail, the two red x's are actually from failing CI caused by rustfmt (PR fix merged already). The release jobs are green :)

**Comment by apoelstra** [comment]
@tcharding oh, cool, I'll rebase to fix that -- but I was referring to the indeteminate future, when we have dependencies set as git tags, etc., and when we get to that point in the upcoming release, we'll have red Xs related to `cargo publish` failing. But I guess it'd be fine to just ignore them.

**Comment by tcharding** [comment]
Oh i get you. For me personally having red sucks because I'm inclined to start ignoring red on other PRs thinking its the same reason - I've done this before and it makes me get sloppy pretty fast.

**Comment by apoelstra** [comment]
Agreed. I'm sure there's some way I can make certain checks have a yellow ! or something, but I can't find it..

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+name: Release
+
+on: [push, pull_request]
```
Comment (sanket1729) [inline .github/workflows/release.yml:3]: Do we need this on both push and pull_request?

Comment (apoelstra) [reply .github/workflows/release.yml:3]: Ah, no, good call. I think I should copy the `on:` field from #1733

**Comment by sanket1729** [comment]
In automation spirit of other ongoing PRs, I think we can have a script check if this PR is a release PR. There are a couple ways of doing this. 
- Check if current PR diff touches the version field in Cargo.toml. (might be annoying to implement)
- fetch the latest version from crates and check our current version against it. (should be easy to check)

For non release PRs, it should return green mark, and for release PRs, it should fail with error.

**Comment by apoelstra** [comment]
> For non release PRs, it should return green mark, and for release PRs, it should fail with error.

I dunno about this. In our last release we had a hiccup because we moved the `bitcoin` crate from the repo root to the `bitcoin/` directory, but then didn't change the `readme` field of the Cargo.toml to be `../README.md` rather than `README.md`.

If we had a check like this, we would've noticed the mistake at the time that we made it, rather than noticing it when we opened the release PR. Obviously either case is better than noticing it after typing `cargo publish` for real :) but I would prefer to know whether each of our crates are in a ready-to-release state.

**Review by tcharding — APPROVED** [review]
ACK 99cb83c3e9a5c98be9289f7d0feab087f4465cef

**Review by Kixunil — APPROVED** [review]
ACK 99cb83c3e9a5c98be9289f7d0feab087f4465cef

**Pull request merged**